crossbeam-channel/src/channel.rs

```diff
@@ -665,6 +665,81 @@ impl<T> Receiver<T> {
         }
     }
 
+    /// Attempts to receive multiple messages from the channel without blocking.
+    ///
+    /// This method will either receive multiple messages from the channel immediately, up to `limit`,
+    /// or return an error if the channel is empty.
+    ///
+    /// If called on a zero-capacity channel, this method will receive a message only if there
+    /// happens to be a send operation on the other side of the channel at the same time.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crossbeam_channel::{unbounded, TryRecvError};
+    ///
+    /// let mut items = Vec::new();
+    ///
+    /// let (s, r) = unbounded();
+    /// assert_eq!(r.try_recv_many(&mut items, 2), Err(TryRecvError::Empty));
+    ///
+    /// s.send(4).unwrap();
+    /// s.send(5).unwrap();
+    /// s.send(6).unwrap();
+    /// s.send(7).unwrap();
+    /// drop(s);
+    ///
+    /// assert_eq!(r.try_recv_many(&mut items, 2), Ok(2));
+    /// assert_eq!(r.try_recv_many(&mut items, 2), Ok(2));
+    /// assert_eq!(r.try_recv_many(&mut items, 2), Err(TryRecvError::Disconnected));
+    /// ```
+    pub fn try_recv_many<E: Extend<T>>(&self, buf: &mut E, limit: usize) -> Result<usize, TryRecvError> {
+        let mut n = 0;
+
+        loop {
+            if n == limit {
+                return Ok(n)
+            }
+
+            let result = match &self.flavor {
+                ReceiverFlavor::Array(chan) => return chan.try_recv_many(buf, limit),
+                ReceiverFlavor::List(chan) => chan.try_recv(),
+                ReceiverFlavor::Zero(chan) => chan.try_recv(),
+                ReceiverFlavor::After(chan) => {
+                    let msg = chan.try_recv();
+                    unsafe {
+                        mem::transmute_copy::<Result<Instant, TryRecvError>, Result<T, TryRecvError>>(
+                            &msg,
+                        )
+                    }
+                }
+                ReceiverFlavor::Tick(chan) => {
+                    let msg = chan.try_recv();
+                    unsafe {
+                        mem::transmute_copy::<Result<Instant, TryRecvError>, Result<T, TryRecvError>>(
+                            &msg,
+                        )
+                    }
+                }
+                ReceiverFlavor::Never(chan) => chan.try_recv(),
+            };
+
+            match result {
+                Ok(msg) => {
+                    buf.extend(Ok::<_, ()>(msg));
```

Comment:
I think it's clearer if you use `std::iter::once`, since `extend` needs an `IntoIterator`.

Reply:
Hmm, nice!  I'll switch it.